cmd/chartsvc/handler.go

```diff
@@ -61,34 +64,97 @@ type rel struct {
 	Links selfLink    `json:"links"`
 }
 
-// listCharts returns a list of charts
-func listCharts(w http.ResponseWriter, req *http.Request) {
+type meta struct {
+	TotalPages int `json:"totalPages"`
+}
+
+// getPageNumberAndSize extracts the page number and size of a request. Default (1, 0) if not set
+func getPageNumberAndSize(req *http.Request) (int, int) {
+	pageNumber := req.FormValue("page")
+	pageSize := req.FormValue("size")
+	numberInt, err := strconv.Atoi(pageNumber)
+	if err != nil || numberInt < 1 {
+		numberInt = 1
+	}
+	sizeInt, _ := strconv.Atoi(pageSize)
+	// Safeguard for negative numbers
+	if sizeInt < 0 {
+		sizeInt = 0
+	}
+	return numberInt, sizeInt
+}
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func getChartList(repo string, pageNumber, pageSize int) (apiListResponse, interface{}, error) {
 	db, closer := dbSession.DB()
 	defer closer()
 	var charts []*models.Chart
-	if err := db.C(chartCollection).Find(nil).Sort("name").All(&charts); err != nil {
+
+	var query datastore.Query
+	c := db.C(chartCollection)
+	if repo != "" {
+		query = c.Find(bson.M{"repo.name": repo})
+	} else {
+		query = c.Find(nil)
+	}
+	query = query.Sort("name")
+
+	// TODO(andresmgot): We should not query the charts just yet in case we paginate the response
+	// but we need to do so since it's not possible to filter & count charts (countDocuments) until
```

Comment:
I don't understand, why can't we use `count`? Some quick testing shows it works:

```
> db.charts.find().count()
345
> db.charts.find({"repo.name": "stable"}).count()
238
> db.charts.find({"repo.name": "bitnami"}).count()
54
```

Reply:
Yes, that's kind of what we are doing here. My point is that when doing `db.charts.find()` you are querying the whole collection while, if using pagination, you are only interested in a certain number of charts:

```
db.charts.find().skip(5).limit(5)
```

That would query just the second page of 5 elements.

ref: https://www.codementor.io/arpitbhayani/fast-and-efficient-pagination-in-mongodb-9095flbqr

Now the goal is to obtain the total amount of pages without querying all the documents in the collection. For doing so, I assume that `db.charts.countDocuments({"repo.name": "stable")}` is faster since it doesn't return the documents but just the sum of elements.

In any case we are not working with huge collections so the optimizations are minimal.

Reply:
Okay, I think I understand what your concern is, but I think you might be misunderstanding how this works.

Firstly, `db.C(chartCollection).Find(...)` only prepares a query, it will not execute it[0]. It gets executed when the returned `Query` Object is called with `All`, `One`, `Count` and other methods that cause an execution.

Secondly, `count` can be optimised in certain cases. As you mentioned, when there isn't a query, it will be performed in constant time[1] by pulling from the collections stats. When there is a query, though, it can be slower. However, indexes in the database would make this faster. I'm not 100% sure about the difference between `countDocuments` and `count`, but from what I can find `countDocuments` is supposedly more accurate in sharded clusters due to using the aggregation API. Also, `count` is apparently deprecated in Mongo v4, but I still think it's worth using it here (we should probably add a comment).

One way to see that `count` is handled a little differently is that it ignores the `skip` and `limit` filters when counting (this is a similar behaviour to `SELECT COUNT(*)` in SQL):

```
> db.charts.find().skip(5).limit(5).count()
345
```

You can use `db.charts.explain()...` to compare how MongoDB handles different queries. Suffice to say, I think we can more than reasonably use `count` to get the total count and then perform the query with skip and limit for retrieving the actual resources. Another very good reason for this is that we avoid loading all charts in memory when we only care about a portion.

> In any case we are not working with huge collections so the optimizations are minimal.

Agreed, which is why we shouldn't need to worry too much about the performance impact of a `count` operation, but fetching everything and doing pagination client-side is an anti-pattern. Databases are designed for this sort of thing, and will be much better at optimising it.

[0] https://godoc.org/github.com/globalsign/mgo#Collection.Find
[1] https://stackoverflow.com/questions/49984134/performance-of-count-for-mongodb